hkmc2/shared/src/test/mlscript-compile/FingerTreeList.mls

```diff
@@ -0,0 +1,253 @@
+  Empty then Single(x)
+  Single(y) then Deep(getNodeSize(y) + getNodeSize(x), [x], Empty, [y])
+  Deep(s, ax, dt, ay) then if (ax.length >= 4)
+    then Deep(s + getNodeSize(x), [x, ax.0], cons(Branch3(
+      getNodeSize(ax.1) + getNodeSize(ax.2) + getNodeSize(ax.3),
+      ax.1, ax.2, ax.3), dt), ay)
+    else Deep(s + getNodeSize(x), [x, ...ax], dt, ay)
+
+// amortized O(1)
+fun (:+) snoc(xs, x) = if safeInput(xs) is
+  Empty then Single(x)
+  Single(y) then Deep(getNodeSize(y) + getNodeSize(x), [y], Empty, [x])
+  Deep(s, ax, dt, ay) then if (ay.length >= 4)
+    then Deep(s + getNodeSize(x), ax, snoc(dt, Branch3(
+      getNodeSize(ay.0) + getNodeSize(ay.1) + getNodeSize(ay.2),
+      ay.0, ay.1, ay.2)), [ay.3, x])
+    else Deep(s + getNodeSize(x), ax, dt, [...ay, x])
+
+fun mk(...args) = args.reduce(
+    (acc, x) => snoc(acc, x),
+    Empty)
+
+fun isEmpty = case
+  Empty then true
+  _ then false
+
+// O(1)
+fun length = case
+  Array as a then a.length
+  Empty then 0
+  Single(x) then getNodeSize(x)
+  Deep(s, _, _, _) then s
+
+// amortized O(1)
+fun popFront(ft) = if safeInput(ft) is
+  Empty then Nil
+  Single(x) then View(x, Empty)
+  Deep(s, [x], dt, ay) then if (popFront(dt)) is
+    Nil then if ay is
+      [y] then View(x, Single(y))
+      [...front, y] then View(x, Deep(s - getNodeSize(x), front, Empty, [y]))
+    View(x', dt') then
+      View(x, Deep(s - getNodeSize(x), arrayOfNode(x'), dt', ay))
+  Deep(s, [x, ...ax], dt, ay) then
+    View(x, Deep(s - getNodeSize(x), ax, dt, ay))
+
+// amortized O(1)
+fun popBack(ft) = if safeInput(ft) is
+  Empty then Nil
+  Single(x) then View(x, Empty)
+  Deep(s, ax, dt, [x]) then if popBack(dt) is
+    Nil then if ax is
+      [y] then View(x, Single(y))
+      [y, ...front] then View(x, Deep(s - getNodeSize(x), [y], Empty, front))
+    View(x', dt') then
+      View(x, Deep(s - getNodeSize(x), ax, dt', arrayOfNode(x')))
+  Deep(s, ax, dt, [...ay, y]) then
+    View(y, Deep(s - getNodeSize(y), ax, dt, ay))
+
+// amortized O(log n)
+fun (++) concat(ft1, ft2) = concatMiddle(safeInput(ft1), [], safeInput(ft2))
+
+// amortized O(log n)
+fun ith(idx) = case
+  Empty then None
+  Single(x) then ithNode(idx, x)
+  Deep(s, ax, dt, ay) and
+    idx >= 0 && idx < s and
+      idx < getAffixSize(ax) then ithAffixElement(idx, ax)
+      idx < getAffixSize(ax) + length(dt) then ith(idx - getAffixSize(ax))(dt)
+      else ithAffixElement(idx - getAffixSize(ax) - length(dt), ay)
+    else None
+
+fun get(idx) = case
+  Array as a then a.at(idx)
+  LFT as lft then if ith(normIdx(idx, length(lft)))(lft) is
+    None then throw "get: index out of bounds"
+    Some(x) then x
+
+fun toArray(ft) = [...ft]
+
+/// amortized O(beg + end)
+fun slice(beg, end)(und) =
+  if und is
+    Array then slice(beg, end)(mk(...und))
+    LFT then if beg < 0 || end < 0 || beg + end > length(und)
+      then throw "slice: index out of bounds"
+      else repeatPopFront(beg)(repeatPopBack(end)(und))
```

Comment:
The argument of the `slice` function seems unconventional. Is there a good reason for choosing end as represent number of element to remove at the end?

Reply:
How about `drop` ?

Reply:
But `drop(n)` is already rather standard (at least in Scala).

Why not `takeAndDrop`?

Reply:
`dropFromEnds`? :smile:

Reply:
Maybe `dropLeftRight`?